src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs

```diff
@@ -0,0 +1,300 @@
+        /// </summary>
+        /// <param name="archiveStream">The stream to write to.</param>
+        /// <param name="archiveFormat">The format of the archive.</param>
+        /// <param name="leaveOpen"><see langword="false"/> to dispose the <paramref name="archiveStream"/> when this instance is disposed; <see langword="true"/> to leave the stream open.</param>
+        /// <remarks><para>If the selected <paramref name="archiveFormat"/> is <see cref="TarFormat.Pax"/>, no Global Extended Attributes entry is written. To write a PAX archive with a Global Extended Attributes entry inserted at the beginning of the archive, use the <see cref="TarWriter(Stream, IEnumerable{KeyValuePair{string, string}}?, bool)"/> constructor instead.</para>
+        /// <para>The recommended format is <see cref="TarFormat.Pax"/> for its flexibility.</para></remarks>
+        /// <exception cref="ArgumentNullException"><paramref name="archiveStream"/> is <see langword="null"/>.</exception>
+        /// <exception cref="IOException"><paramref name="archiveStream"/> is unwritable.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="archiveFormat"/> is either <see cref="TarFormat.Unknown"/>, or not one of the other enum values.</exception>
+        public TarWriter(Stream archiveStream, TarFormat archiveFormat, bool leaveOpen = false)
+        {
+            ArgumentNullException.ThrowIfNull(archiveStream);
+
+            if (!archiveStream.CanWrite)
+            {
+                throw new IOException(SR.IO_NotSupported_UnwritableStream);
+            }
+
+            if (archiveFormat is not TarFormat.V7 and not TarFormat.Ustar and not TarFormat.Pax and not TarFormat.Gnu)
+            {
+                throw new ArgumentOutOfRangeException(nameof(archiveFormat));
+            }
+
+            _archiveStream = archiveStream;
+            Format = archiveFormat;
+            _leaveOpen = leaveOpen;
+            _isDisposed = false;
+            _wroteEntries = false;
+            _wroteGEA = false;
+            _globalExtendedAttributes = null;
+        }
+
+        /// <summary>
+        /// The format of the archive.
+        /// </summary>
+        public TarFormat Format { get; private set; }
+
+        /// <summary>
+        /// Disposes the current <see cref="TarWriter"/> instance, and closes the archive stream if the <c>leaveOpen</c> argument was set to <see langword="false"/> in the constructor.
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Asynchronously disposes the current <see cref="TarWriter"/> instance, and closes the archive stream if the <c>leaveOpen</c> argument was set to <see langword="false"/> in the constructor.
+        /// </summary>
+        public ValueTask DisposeAsync()
+        {
+            throw new NotImplementedException();
```

Comment:
?? Does this need a TODO?

Reply:
I'll address all the async methods in the next PR. I will add a final comment in this PR with all the pending tasks.

Reply:
Would it make more sense to only add the APIs once they are implemented? That way someone doesn't try to use these APIs only to get NIE at runtime?

Reply:
Yeah, makes sense. I'll remove them.

Reply:
Yes, please remove them prior to merging if they're not yet implemented.